### PR TITLE
[thread-cert] refactor case 5.5.7 using pktverify

### PR DESCRIPTION
**Pull request by FiveDimensions**

### CODE REVIEW

File: tests/scripts/thread-cert/Cert_5_5_07_SplitMergeThreeWay.py [hunk 1]
```diff
@@ -111,6 +117,86 @@ def test(self):
...
+        p.filter_mle_cmd(MLE_PARENT_RESPONSE).must_not_next()
+
+        # Step3: The Leader MUST stop sending MLE advertisements.
+        p = leader_pkts.range(lreset_start, lreset_stop)
+        p.filter_mle_cmd(MLE_ADVERTISEMENT).must_not_next()
+
+        # Step 8: Router_1 take over leader role of a new Partition and
+        # begin transmitting MLE Advertisements
+        p = _router1_pkts.copy().filter_mle_cmd(MLE_ADVERTISEMENT).must_next()
+        self.assertTrue([0, 11, 9] == p.mle.tlv.type)
```
Comment (simonlingoogle) [inline tests/scripts/thread-cert/Cert_5_5_07_SplitMergeThreeWay.py:179]: convert to sets

File: tests/scripts/thread-cert/Cert_5_5_07_SplitMergeThreeWay.py [hunk 2]
```diff
@@ -111,6 +117,86 @@ def test(self):
...
+
+        # Step 9: The Leader MUST send properly formatted MLE Parent
+        # Requests to the All-Routers multicast address
+        p = _lpkts.filter_mle_cmd(MLE_PARENT_REQUEST).must_next()
+        self.assertTrue([1, 3, 14, 18] == p.mle.tlv.type)
```
Comment (simonlingoogle) [inline tests/scripts/thread-cert/Cert_5_5_07_SplitMergeThreeWay.py:184]: convert to sets

File: tests/scripts/thread-cert/Cert_5_5_07_SplitMergeThreeWay.py [hunk 3]
```diff
@@ -111,6 +117,86 @@ def test(self):
...
+
+        # Step 10: Router_1 MUST send an MLE Parent Response
+        p = _router1_pkts.filter_mle_cmd(MLE_PARENT_RESPONSE).must_next()
+        self.assertTrue(set([0, 11, 5, 4, 3, 16, 15, 18]) < set(p.mle.tlv.type))
```
Comment (simonlingoogle) [inline tests/scripts/thread-cert/Cert_5_5_07_SplitMergeThreeWay.py:188]:
```suggestion
        self.assertTrue({0, 11, 5, 4, 3, 16, 15, 18} <= set(p.mle.tlv.type))
```
btw, use `must_verify`

File: tests/scripts/thread-cert/Cert_5_5_07_SplitMergeThreeWay.py [hunk 4]
```diff
@@ -111,6 +117,86 @@ def test(self):
...
+
+        # Step 11: Leader send MLE Child ID Request
+        p = _lpkts.filter_mle_cmd(MLE_CHILD_ID_REQUEST).must_next()
+        self.assertTrue(
+            set([4, 5, 1, 2, 18, 13, 10, 12, 9, 22]) < set(p.mle.tlv.type))
```
Comment (simonlingoogle) [inline tests/scripts/thread-cert/Cert_5_5_07_SplitMergeThreeWay.py:193]: Ditto

File: tests/scripts/thread-cert/Cert_5_5_07_SplitMergeThreeWay.py [hunk 5]
```diff
@@ -111,6 +117,86 @@ def test(self):
...
+        leader_pkts.filter_mle_cmd(MLE_CHILD_ID_RESPONSE).must_next()
+        _start_idx = router3_pkts.index
+        _lpkts = leader_pkts.range(_start_idx)
+        _router1_pkts = router1_pkts.range(_start_idx)
+
+        # Step 2: The Leader and Router_1 MUST send properly formatted MLE Advertisements
+        leader_pkts.filter_mle_cmd(MLE_CHILD_ID_RESPONSE).must_next()
+        _lpkts = leader_pkts.range(leader_pkts.index)
+        p = _lpkts.filter_mle_cmd(MLE_ADVERTISEMENT).must_next()
+        self.assertTrue([0, 11, 9] == p.mle.tlv.type)
```
[the remark this answers is not in the document]
Comment (FiveDimensions) [reply tests/scripts/thread-cert/Cert_5_5_07_SplitMergeThreeWay.py:142]: @simonlingoogle yes, different order in lists will cause inequality

**Comment by jwhui** [comment]
@FiveDimensions , please help resolve conflicts.

**Comment by FiveDimensions** [comment]
@jwhui sure, rebase to master

File: tests/scripts/thread-cert/Cert_5_5_07_SplitMergeThreeWay.py [hunk 6]
```diff
@@ -111,6 +117,74 @@ def test(self):
...
+
+        # Step 2: The Leader and Router_1 MUST send properly formatted MLE Advertisements
+        leader_pkts.filter_mle_cmd(MLE_CHILD_ID_RESPONSE).must_next()
+        _lpkts = leader_pkts.copy()
+        _lpkts.filter_mle_cmd(MLE_ADVERTISEMENT).must_next().must_verify(lambda p: {0, 11, 9} == set(p.mle.tlv.type))
+
+        router1_pkts.range(leader_pkts.index).filter_mle_cmd(MLE_ADVERTISEMENT).must_next().must_verify(
+            lambda p: {0, 11, 9} == set(p.mle.tlv.type))
+
+        # Step4: Each router forms a partition with the
```
Comment (jwhui) [inline tests/scripts/thread-cert/Cert_5_5_07_SplitMergeThreeWay.py:141]:
```suggestion
        # Step 4: Each router forms a partition with the
```

File: tests/scripts/thread-cert/Cert_5_5_07_SplitMergeThreeWay.py [hunk 7]
```diff
@@ -111,6 +117,74 @@ def test(self):
...
+        # lowest possible partition ID
+        # Step 5: Router_1 MUST send MLE Parent Requests
```
Comment (jwhui) [inline tests/scripts/thread-cert/Cert_5_5_07_SplitMergeThreeWay.py:143]: nit: why is the line wrap length here different than other lines?

File: tests/scripts/thread-cert/Cert_5_5_07_SplitMergeThreeWay.py [hunk 8]
```diff
@@ -111,6 +117,74 @@ def test(self):
...
+        # within range by sending a MLE Parent Request.
+        router1_pkts.filter_mle_cmd(MLE_PARENT_REQUEST).must_next().must_verify(
+            lambda p: {1, 3, 14, 18} == set(p.mle.tlv.type))
+        lreset_stop = router1_pkts.index
+
+        # Step 6: The Leader does not respond to the Parent Requests
+        p = leader_pkts.range(lreset_start, lreset_stop)
+        p.filter_mle_cmd(MLE_PARENT_RESPONSE).must_not_next()
+
+        # Step3: The Leader MUST stop sending MLE advertisements.
```
Comment (jwhui) [inline tests/scripts/thread-cert/Cert_5_5_07_SplitMergeThreeWay.py:161]:
```suggestion
        # Step 7: The Leader MUST stop sending MLE advertisements.
```

**Comment by FiveDimensions** [comment]
@jwhui the reason of Step 3 and Step 6 disorders is to verify if leader does not sent out the MLE advertisements and Parent Requests between step2 to step 7

File: tests/scripts/thread-cert/Cert_5_5_07_SplitMergeThreeWay.py [hunk 9]
```diff
@@ -111,6 +117,71 @@ def test(self):
...
+
+        # Step 4: Each router forms a partition with the lowest possible partition ID
+        # Step 5: Router_1 MUST send MLE Parent Requests and MUST make two separate attempts
+        router1_pkts.filter_mle_cmd(MLE_PARENT_REQUEST).must_next().must_verify(lambda p: {1, 3, 14, 18} == set(
+            p.mle.tlv.type) and p.mle.tlv.scan_mask.r == 1 and p.mle.tlv.scan_mask.e == 1)
+        lreset_start = router1_pkts.index
+        router1_pkts.filter_mle_cmd(MLE_PARENT_REQUEST).must_next().must_verify(lambda p: {1, 3, 14, 18} == set(
+            p.mle.tlv.type) and p.mle.tlv.scan_mask.r == 1 and p.mle.tlv.scan_mask.e == 1)
+
+        # Step 7:Router_1 MUST attempt to attach to any other Partition
```
Comment (jwhui) [inline tests/scripts/thread-cert/Cert_5_5_07_SplitMergeThreeWay.py:149]:
```suggestion
        # Step 7: Router_1 MUST attempt to attach to any other Partition
```

File: tests/scripts/thread-cert/Cert_5_5_07_SplitMergeThreeWay.py [hunk 10]
```diff
@@ -111,6 +117,71 @@ def test(self):
...
+
+        # Step 10: Router_1 MUST send an MLE Parent Response
+        router1_pkts.filter_mle_cmd(MLE_PARENT_RESPONSE).must_next().must_verify(
+            lambda p: {0, 11, 5, 4, 3, 16, 15, 18} <= set(p.mle.tlv.type))
+
+        # Step 11: Leader send MLE Child ID Request
+        _lpkts.filter_mle_cmd(MLE_CHILD_ID_REQUEST).must_next().must_verify(
+            lambda p: {4, 5, 1, 2, 18, 13, 10, 12, 9, 22} <= set(p.mle.tlv.type))
+
+        #Step 12: DUT (Router or Leader) MUST respond with a ICMPv6 Echo Reply
```
Comment (jwhui) [inline tests/scripts/thread-cert/Cert_5_5_07_SplitMergeThreeWay.py:180]:
```suggestion
        # Step 12: DUT (Router or Leader) MUST respond with a ICMPv6 Echo Reply
```

File: tests/scripts/thread-cert/Cert_5_5_07_SplitMergeThreeWay.py [hunk 11]
```diff
@@ -111,6 +117,71 @@ def test(self):
...
+        router2_pkts = pkts.filter_wpan_src64(ROUTER_2)
+        router3_pkts = pkts.filter_wpan_src64(ROUTER_3)
+
+        # Step 2: The Leader and Router_1 MUST send properly formatted MLE Advertisements
+        leader_pkts.filter_mle_cmd(MLE_CHILD_ID_RESPONSE).must_next()
+        _lpkts = leader_pkts.copy()
+        _lpkts.filter_mle_cmd(MLE_ADVERTISEMENT).must_next().must_verify(lambda p: {0, 11, 9} == set(p.mle.tlv.type))
+
+        router1_pkts.range(leader_pkts.index).filter_mle_cmd(MLE_ADVERTISEMENT).must_next().must_verify(
+            lambda p: {0, 11, 9} == set(p.mle.tlv.type))
```
Comment (simonlingoogle) [inline tests/scripts/thread-cert/Cert_5_5_07_SplitMergeThreeWay.py:139]: nit: define mle tlv type constants

File: tests/scripts/thread-cert/Cert_5_5_07_SplitMergeThreeWay.py [hunk 12]
```diff
@@ -111,6 +117,71 @@ def test(self):
...
+
+        # Step 4: Each router forms a partition with the lowest possible partition ID
+        # Step 5: Router_1 MUST send MLE Parent Requests and MUST make two separate attempts
+        router1_pkts.filter_mle_cmd(MLE_PARENT_REQUEST).must_next().must_verify(lambda p: {1, 3, 14, 18} == set(
```
Comment (simonlingoogle) [inline tests/scripts/thread-cert/Cert_5_5_07_SplitMergeThreeWay.py:143]: nit: define mle tlv type constants

File: tests/scripts/thread-cert/Cert_5_5_07_SplitMergeThreeWay.py [hunk 13]
```diff
@@ -111,6 +117,71 @@ def test(self):
...
+            p.mle.tlv.type) and p.mle.tlv.scan_mask.r == 1 and p.mle.tlv.scan_mask.e == 1)
+        lreset_start = router1_pkts.index
+        router1_pkts.filter_mle_cmd(MLE_PARENT_REQUEST).must_next().must_verify(lambda p: {1, 3, 14, 18} == set(
```
Comment (simonlingoogle) [inline tests/scripts/thread-cert/Cert_5_5_07_SplitMergeThreeWay.py:146]: nit: define mle tlv type constants

File: tests/scripts/thread-cert/Cert_5_5_07_SplitMergeThreeWay.py [hunk 14]
```diff
@@ -111,6 +117,71 @@ def test(self):
...
+            p.mle.tlv.type) and p.mle.tlv.scan_mask.r == 1 and p.mle.tlv.scan_mask.e == 1)
+
+        # Step 7:Router_1 MUST attempt to attach to any other Partition
+        # within range by sending a MLE Parent Request.
+        router1_pkts.filter_mle_cmd(MLE_PARENT_REQUEST).must_next().must_verify(
```
Comment (simonlingoogle) [inline tests/scripts/thread-cert/Cert_5_5_07_SplitMergeThreeWay.py:151]: nit: define mle tlv type constants

File: tests/scripts/thread-cert/Cert_5_5_07_SplitMergeThreeWay.py [hunk 15]
```diff
@@ -111,6 +117,71 @@ def test(self):
...
+        p = leader_pkts.range(lreset_start, lreset_stop)
+        p.filter_mle_cmd(MLE_ADVERTISEMENT).must_not_next()
+
+        # Step 6: The Leader does not respond to the Parent Requests
+        p = leader_pkts.range(lreset_start, lreset_stop)
+        p.filter_mle_cmd(MLE_PARENT_RESPONSE).must_not_next()
+
+        # Step 8: Router_1 take over leader role of a new Partition and begin transmitting MLE Advertisements
+        router1_pkts.copy().filter_mle_cmd(MLE_ADVERTISEMENT).must_next().must_verify(
+            lambda p: {0, 11, 9} == set(p.mle.tlv.type))
```
Comment (simonlingoogle) [inline tests/scripts/thread-cert/Cert_5_5_07_SplitMergeThreeWay.py:165]: nit: define mle tlv type constants

File: tests/scripts/thread-cert/Cert_5_5_07_SplitMergeThreeWay.py [hunk 16]
```diff
@@ -111,6 +117,71 @@ def test(self):
...
+
+        # Step 9: The Leader MUST send properly formatted MLE Parent Requests to the
+        # All-Routers multicast address
+        _lpkts.filter_mle_cmd(MLE_PARENT_REQUEST).must_next().must_verify(
+            lambda p: {1, 3, 14, 18} == set(p.mle.tlv.type))
```
Comment (simonlingoogle) [inline tests/scripts/thread-cert/Cert_5_5_07_SplitMergeThreeWay.py:170]: nit: define mle tlv type constants

File: tests/scripts/thread-cert/Cert_5_5_07_SplitMergeThreeWay.py [hunk 17]
```diff
@@ -111,6 +117,71 @@ def test(self):
...
+
+        # Step 10: Router_1 MUST send an MLE Parent Response
+        router1_pkts.filter_mle_cmd(MLE_PARENT_RESPONSE).must_next().must_verify(
+            lambda p: {0, 11, 5, 4, 3, 16, 15, 18} <= set(p.mle.tlv.type))
+
+        # Step 11: Leader send MLE Child ID Request
+        _lpkts.filter_mle_cmd(MLE_CHILD_ID_REQUEST).must_next().must_verify(
+            lambda p: {4, 5, 1, 2, 18, 13, 10, 12, 9, 22} <= set(p.mle.tlv.type))
```
Comment (simonlingoogle) [inline tests/scripts/thread-cert/Cert_5_5_07_SplitMergeThreeWay.py:178]: nit: define mle tlv type constants

File: tests/scripts/thread-cert/Cert_5_5_07_SplitMergeThreeWay.py [hunk 18]
```diff
@@ -111,6 +117,71 @@ def test(self):
...
+            lambda p: {0, 11, 9} == set(p.mle.tlv.type))
+
+        # Step 9: The Leader MUST send properly formatted MLE Parent Requests to the
+        # All-Routers multicast address
+        _lpkts.filter_mle_cmd(MLE_PARENT_REQUEST).must_next().must_verify(
+            lambda p: {1, 3, 14, 18} == set(p.mle.tlv.type))
+
+        # Step 10: Router_1 MUST send an MLE Parent Response
+        router1_pkts.filter_mle_cmd(MLE_PARENT_RESPONSE).must_next().must_verify(
+            lambda p: {0, 11, 5, 4, 3, 16, 15, 18} <= set(p.mle.tlv.type))
```
Comment (simonlingoogle) [inline tests/scripts/thread-cert/Cert_5_5_07_SplitMergeThreeWay.py:174]: nit: define mle tlv type constants

**Comment by jwhui** [comment]
@FiveDimensions , can you help resolve the comments above?

File: tests/scripts/thread-cert/Cert_5_5_07_SplitMergeThreeWay.py [hunk 19]
```diff
@@ -111,6 +117,80 @@ def test(self):
...
+
+        # Step 7: Router_1 MUST attempt to attach to any other Partition
+        # within range by sending a MLE Parent Request.
+        router1_pkts.filter_mle_cmd(MLE_PARENT_REQUEST).must_next().must_verify(
+            lambda p: {MODE_TLV, CHALLENGE_TLV, SCAN_MASK_TLV, VERSION_TLV} == set(p.mle.tlv.type))
+        lreset_stop = router1_pkts.index
+
+        # Step 3: The Leader MUST stop sending MLE advertisements.
+        p = leader_pkts.range(lreset_start, lreset_stop)
+        p.filter_mle_cmd(MLE_ADVERTISEMENT).must_not_next()
```
Comment (simonlingoogle) [inline tests/scripts/thread-cert/Cert_5_5_07_SplitMergeThreeWay.py:160]: We usually use `p` for a packet. We never use `p` for a `PacketFilter`.

Seems `p` is not necessary.
```suggestion
        leader_pkts.range(lreset_start, lreset_stop).filter_mle_cmd(MLE_ADVERTISEMENT).must_not_next()
```

File: tests/scripts/thread-cert/Cert_5_5_07_SplitMergeThreeWay.py [hunk 20]
```diff
@@ -111,6 +117,80 @@ def test(self):
...
+
+        # Step 6: The Leader does not respond to the Parent Requests
+        p = leader_pkts.range(lreset_start, lreset_stop)
```
Comment (simonlingoogle) [inline tests/scripts/thread-cert/Cert_5_5_07_SplitMergeThreeWay.py:163]: ditto

**Review by simonlingoogle — APPROVED** [review]
LGTM :+1: 
with a nit

**Pull request merged**